[CPU] Broadcast and Tile optimized

antonvor opened the pull request:
### Issues: 29769, 35807, 33187, 33233

### This PR includes:
- Broadcast and Tile call either optimized implementation or ngraph reference on CPU
- Disabled "Broadcast to Tile" and "Tile to TileIE" conversions on CPU
- Added Broadcast and Tile shared tests + CPU specific tests
- Supported blocked layouts, some precisions and arbitrary number of dimensions

ilyachur commented on inference-engine/src/legacy_api/src/ie_cnn_layer_builder_ngraph.cpp:
```diff
@@ -1548,6 +1553,17 @@ CNNLayer::Ptr NodeConverter<ngraph::op::Eltwise>::createLayer(const std::shared_
 }
 
+template <>
+CNNLayer::Ptr NodeConverter<ngraph::op::Tile>::createLayer(const std::shared_ptr<ngraph::Node>& layer) const {
```
Please use visitor_api for tile layer instead of NodeConverter

antonvor replied:
Thanks for your comment, changed

antonvor commented:
@openvinotoolkit/openvino-ie-cpu-developers take a look please
@dmitry-gorokhov could you please review?

dmitry-gorokhov commented:
@nshchego  Please review the PR

nshchego commented on inference-engine/src/mkldnn_plugin/nodes/broadcast.cpp:
```diff
@@ -1,135 +1,186 @@
+//// SPDX-License-Identifier: Apache-2.0
+////
+//
+//#include "base.hpp"
```
We should not keep commented code in repository.

antonvor replied:
thanks, removed

nshchego commented on inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.h:
```diff
@@ -0,0 +1,32 @@
+#include <ngraph/runtime/host_tensor.hpp>
+#include "mkldnn_node.h"
+
+using namespace InferenceEngine;
```
Don't use 'using namespace' in headers

antonvor replied:
fixed

nshchego commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_broadcast_node.cpp:
```diff
@@ -0,0 +1,225 @@
+        LayerConfig config;
+        DataConfig dataConfig, shapeConfig;
+        Precision dataPrecision = getCnnLayer()->outData[0]->getTensorDesc().getPrecision();
+        const SizeVector& data_dims = getCnnLayer()->insData[0].lock()->getTensorDesc().getDims();
```
insData[0] is weak_ptr and must be validated

antonvor replied:
fixed, added check

nshchego commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_broadcast_node.cpp:
```diff
@@ -0,0 +1,225 @@
+        dataConfig.desc = TensorDesc(dataPrecision, data_dims,
+                                     getCnnLayer()->insData[0].lock()->getTensorDesc().getLayout());
+        config.inConfs.push_back(dataConfig);
+        SizeVector shape_dims = getCnnLayer()->insData[1].lock()->getTensorDesc().getDims();
```
Use const reference here

antonvor replied:
fixed

nshchego commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_broadcast_node.cpp:
```diff
@@ -0,0 +1,225 @@
+        if (!srcStrides.size())
+            srcStrides = SizeVector(1, 1);
+
+        if (dst_dims.size() != shape_size) {
```
Seems it could be validated on node creation.

antonvor replied:
It was a simple code transfer. Now it is validated earlier.

nshchego commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_tile_node.cpp:
```diff
@@ -3,74 +3,91 @@
+    } else {
+        InferenceEngine::Precision precision = getCnnLayer()->insData[0].lock()->getPrecision();
+        if (precision != InferenceEngine::Precision::FP32)
+            precision = InferenceEngine::Precision::FP32;
```
Why we cannot set FP32 directly for input and output without retrieving and check precision from Data?

antonvor replied:
You're right, fixed

nshchego commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_tile_node.cpp:
```diff
@@ -3,74 +3,91 @@
+    if (tile) {
+        supportedPrimitiveDescriptors = getSupportedConfigs(this, optimizedCase, repeats);
+    } else {
+        InferenceEngine::Precision precision = getCnnLayer()->insData[0].lock()->getPrecision();
```
insData[0].lock() must be validated.

antonvor replied:
fixed

antonvor commented on inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp:
```diff
@@ -0,0 +1,192 @@
+    uint8_t *srcData = reinterpret_cast<uint8_t*>(node->getParentEdgeAt(0)->getMemory().GetData()) +
+                       node->getParentEdgeAt(0)->getMemory().GetDescriptor().data.layout_desc.blocking.offset_padding;
+    uint8_t* dstData = reinterpret_cast<uint8_t*>(node->getChildEdgeAt(0)->getMemory().GetData()) +
+                       node->getChildEdgeAt(0)->getMemory().GetDescriptor().data.layout_desc.blocking.offset_padding;
```
Thanks a lot for your comment! This place could cause bugs in the future

antonvor commented:
@dmitry-gorokhov could you please finally review?

dmitry-gorokhov commented:
> @dmitry-gorokhov could you please finally review?

@antonvor  I don't see approval from @nshchego.

dmitry-gorokhov commented on inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/broadcast.cpp:
```diff
@@ -0,0 +1,89 @@
+        InferenceEngine::Precision::I8,
+        InferenceEngine::Precision::U8,
+        InferenceEngine::Precision::I32,
+        InferenceEngine::Precision::FP32
```
Lets add smt with size == 2 bytes

antonvor replied:
added

dmitry-gorokhov commented on inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/broadcast.cpp:
```diff
@@ -0,0 +1,89 @@
+                                        ::testing::Values(std::vector<size_t>{16, 50, 50},
+                                                          std::vector<size_t>{1, 16, 50, 50},
+                                                          std::vector<size_t>{1, 1, 16, 50, 50},
+                                                          std::vector<size_t>{2, 16, 1, 5}),
```
Can we cover 1D and 2D inputs as well?

antonvor replied:
yes, added 1D and 2D tests

antonvor commented:
@itikhono Could you take a look please? I changed the logic of Tile and Broadcast nodes in the CPU plugin and removed some conversions, can I fix OpsShouldBeConvertedToIERepresentation test this way?

rblaczkowski commented:
Please merge master to your branch to fetch the newest OpenVino ONNX CI state.

dmitry-gorokhov commented on inference-engine/src/legacy_api/src/transformations/convert_opset1_to_legacy/convert_tile_to_ie_tile.cpp:
```diff
@@ -19,9 +19,9 @@ ngraph::pass::ConvertTileToLegacyMatcher::ConvertTileToLegacyMatcher() {
+    ngraph::matcher_pass_callback callback = [this](pattern::Matcher& m) {
         auto tile = std::dynamic_pointer_cast<ngraph::opset1::Tile> (m.get_match_root());
-        if (!tile) {
+        if (!tile || m_transformation_callback(tile)) {
```
Do we use this callback?

antonvor replied:
No, you're right, fixed

dmitry-gorokhov commented:
@ilyachur  May I ask you to take a look on changes in Ngraph core and conversion to legacy parts.

iefode commented on inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/broadcast.cpp:
```diff
@@ -13,162 +13,161 @@ namespace {
+        ::testing::Values(CommonTestUtils::DEVICE_CPU)
+);
+
+INSTANTIATE_TEST_CASE_P(TestNumpyBroadcast1D, BroadcastLayerTest, numpyBroadcast1DInputParams, BroadcastLayerTest::getTestCaseName);
```
Please. add `smoke` prefix to add these tests to the pre-commit

antonvor replied:
added

ilyachur commented on inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:
```diff
@@ -921,6 +928,7 @@ void convertFunctionToICNNNetwork(const std::shared_ptr<const ::ngraph::Function
         };
         const static std::vector<std::shared_ptr<Builder::INodeConverter>> convertors = {
                 std::make_shared<Builder::NodeConverter<::ngraph::op::v1::AvgPool>>(),
+                std::make_shared<Builder::NodeConverter<::ngraph::op::v1::Broadcast>>(),
```
Please use visitor API instead of node converter.

antonvor replied:
Thanks! Changed

ilyachur commented on inference-engine/src/legacy_api/src/ie_cnn_layer_builder_ngraph.cpp:
```diff
@@ -639,6 +639,17 @@ CNNLayer::Ptr NodeConverter<ngraph::op::v1::Add>::createLayer(const std::shared_
 }
 
+template <>
+CNNLayer::Ptr NodeConverter<ngraph::op::v1::Broadcast>::createLayer(const std::shared_ptr<ngraph::Node>& layer) const {
```
We don't need this converter if we use Visitor API.

antonvor commented:
@dmitry-gorokhov could you please review?

dmitry-gorokhov commented:
new PR: https://github.com/openvinotoolkit/openvino/pull/8041
